src/GafferScene/Shader.cpp

```diff
@@ -474,6 +474,12 @@ class Shader::NetworkBuilder
 			}
 
 			const Shader *effectiveShader = static_cast<const Shader *>( effectiveParameter->node() );
+
+			// TODO - talk to John about isInputParameter vs !isOutputParameter.
+			// Based on asserts elsewhere, it seems we can assume they are synonyms, but !isOutput would
+			// seem to express the intent more clearly - the else branch assumes that we've found an
+			// output parameter.  This might actually be true for all uses of isInput... maybe it shouldn't exist?
```

Comment:
`isInputParameter()` is only a synonym of `!isOutputParameter()` if you've already verified that what you have is a parameter, not just a random plug. This comes into play at least once, at L258, where `effectiveParameter()` ignores connections from non-parameter plugs. I haven't checked for other examples, but I think both `isInputParameter()` and `isOutputParameter()` are best left as-is, so they do what they say on the tin, without relying on any additional assumptions.